packages/react-i18n/src/index.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * WordPress dependencies
+ */
+import React, {
+	createContext,
+	useContext,
+	useEffect,
+	useMemo,
+	useState,
+	useCallback,
+} from '@wordpress/element';
+import { createI18n, getLocaleData } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
+import { addQueryArgs } from '@wordpress/url';
+
+const SCRIPT_HANDLES = [];
+
+const Context = createContext( null );
+
+export function LocaleProvider( { locale: initialLocale, domain, children } ) {
+	const [ locale, setLocale ] = useState( initialLocale );
+	const [ localeMap, setLocaleMap ] = useState( () => {
+		return new Map().set(
+			initialLocale,
+			createI18n( getLocaleData( domain ), domain )
+		);
+	} );
+
+	const switchToLocale = useCallback(
+		( newLocale ) => {
+			if ( ! localeMap.get( newLocale ) ) {
+				setLocaleMap(
+					new Map(
+						localeMap.set( newLocale, createI18n( {}, domain ) )
+					)
+				);
+			}
+			setLocale( newLocale );
+		},
+		[ localeMap ]
+	);
+
+	// Adds new locale data (= translations) and updates the map to force a re-render.
+	const setLocaleData = useCallback(
+		( data ) => {
+			localeMap.get( locale ).setLocaleData( data );
+			setLocaleMap(
+				new Map( localeMap.set( locale, localeMap.get( locale ) ) )
+			);
+		},
+		[ locale, localeMap ]
+	);
+
+	// TODO: Do some caching.
+	useEffect( () => {
+		async function fetchTranslations() {
+			const result = await apiFetch( {
+				path: addQueryArgs( `/__experimental/translations`, {
+					handles: SCRIPT_HANDLES,
+					locale,
+				} ),
+			} );
+
+			for ( const translations of Object.values( result ) ) {
+				const localeData =
+					translations.locale_data[ domain ] ||
+					translations.locale_data.messages;
+				localeData[ '' ].domain = domain;
+
+				localeMap.get( locale ).setLocaleData( localeData );
+			}
+
+			setLocaleMap(
+				new Map( localeMap.set( locale, localeMap.get( locale ) ) )
+			);
+		}
+
+		if ( 'en_US' !== locale ) {
+			fetchTranslations();
+		}
+	}, [ locale, domain ] );
+
+	const value = useMemo( () => {
+		const { __, _x, _n, _nx, isRtl } = localeMap?.get( locale ) || {};
+
+		return {
+			locale,
+			__,
```

Comment:
It's very useful to bind the `__` functions to create unique instances on every context change: `return { __: __.bind() }`. Without this, after `setLocaleData()` the `i18n` object is still the same and therefore the `__` function is still the same. Even though the translation data have changed and components should be updated.

If the `__` value doesn't change, usages like this will fail to rerender and update:
```jsx
const Heading = React.memo( ( { __ } ) => {
  return <h1>{ __( 'Warning' ) }</h1>;
} );

const Box = ( text ) => {
  const { __ } = useTranslate();
  return (
    <div>
      <Heading __={ __ } />
      <p>{ __( 'Hello' ) }</p>
    </div>
  );
}
```
A component that uses `React.memo` or `React.PureComponent` will rerender only when props change, and for `<Heading />` and `setLocaleData()` they don't.

This scenario will happen any time there is a `withTranslate` higher-order component, which passes `__` down as a prop, and a Redux-`connect`-ed component, which by default behaves as pure.

Reply:
Ah, good call!